src/librbd/CopyupRequest.cc

```diff
@@ -259,36 +282,74 @@ void CopyupRequest::remove_from_list()
   m_ictx->copyup_list.erase(it);
 }
 
-bool CopyupRequest::send_object_map() {
+bool CopyupRequest::send_object_map_head() {
+  CephContext *cct = m_ictx->cct;
+  ldout(cct, 20) << __func__ << " " << this << dendl;
+
+  m_state = STATE_OBJECT_MAP_HEAD;
+
   {
+    RWLock::RLocker owner_locker(m_ictx->owner_lock);
     RWLock::RLocker snap_locker(m_ictx->snap_lock);
     if (m_ictx->object_map != nullptr) {
       bool copy_on_read = m_pending_requests.empty();
       assert(m_ictx->exclusive_lock->is_lock_owner());
 
       RWLock::WLocker object_map_locker(m_ictx->object_map_lock);
-      if (copy_on_read &&
-          (*m_ictx->object_map)[m_object_no] != OBJECT_EXISTS) {
-        // CoW already updates the HEAD object map
-        m_snap_ids.push_back(CEPH_NOSNAP);
-      }
       if (!m_ictx->snaps.empty()) {
         m_snap_ids.insert(m_snap_ids.end(), m_ictx->snaps.begin(),
                           m_ictx->snaps.end());
       }
+      if (copy_on_read &&
+          (*m_ictx->object_map)[m_object_no] != OBJECT_EXISTS) {
+        m_snap_ids.insert(m_snap_ids.begin(), CEPH_NOSNAP);
+        object_map_locker.unlock();
+        snap_locker.unlock();
+        owner_locker.unlock();
+        return send_object_map_snap();
```

Comment:
State name doesn't really align here for the copy-on-read case since it is also updating the HEAD revision. Maybe keep the original name?